### ISSUE TITLE
OOM in forward pass of ESREncoder

### ISSUE THREAD

**Issue by man-sean** [leverxgroup/esrgan#5]
I get an OOM error when forwarding an input of size `16*3*128*128` (as used by the paper) through `ESREncoder` with 16 RRDB blocks (as used by the paper), to reproduce:

```python
import torch
from esrgan.model.module.esrnet import ESREncoder

esr = ESREncoder(num_basic_blocks=16).to('cuda')
input_ = torch.randn((16,3,128,128)).to('cuda')
esr(input_)
```

I'm running on 1080ti with 11GB, which is very similar in terms of memory to Titan Xp.
Environment wise I cloned the repo into a Conda environment with `torch 1.9.0+cu111` and `CUDA 11.2`.

What am I missing?
Thanks!

**Comment by man-sean** [comment]
I installed a new environment following the instruction in the README.

Running `catalyst-dl run -C esrgan/experiment/config_supervised.yml --benchmark` succeeds (and only 7GB are consumed).

But still running the above snippet result in OOM error.

At the beginning of `catalyst-dl` I got the following warning:
```
~/miniconda3/envs/esrgan-debug/lib/python3.8/site-packages/piq/utils/common.py:15: UserWarning:
Expected input tensor <class 'torch.FloatTensor'>, got torch.cuda.HalfTensor.
```
Does the training uses mixed-precision or some other memory optimization?
I didn't see any clear indication in the config / code.

**Comment by bagxi** [comment]
Hi,

But still running the above snippet result in OOM error. - Could you please try smaller batch size, for example `torch.randn((2,3,128,128))`?

Does the training uses mixed-precision or some other memory optimization? - Yes, by default AMP (fp16) is used.

**Comment by man-sean** [comment]
Thanks for the reply ☺️

Smaller batch size indeed works (batch of 2 takes 5.6GB).

If AMP is used this makes sense. Where in the code this flag was used? I'm not familiar with Catalyst so I must have missed that.

Thanks again!

**Comment by man-sean** [comment]
To check if AMP solves the problem I ran the following:

```python
import torch
from torch.cuda.amp import autocast
from esrgan.model.module.esrnet import ESREncoder

esr = ESREncoder(num_basic_blocks=16).to('cuda')
input_ = torch.randn((16,3,128,128)).to('cuda')
with autocast():
    esr(input_)
```

With BS=2 I indeed see the memory usage reduces from 5.6GB to ~3GB, but BS=16 still throws OOM.
Are there any other optimization used? Did I misconfigured AMP?

**Comment by bagxi** [comment]
Actually, AMP used only during the training phase: [AMPOptimizerCallback](https://github.com/leverxgroup/esrgan/blob/master/experiment/config_supervised.yml#L162). Btw, you can replace it with [OptimizerCallback](https://catalyst-team.github.io/catalyst/v20.10/api/callbacks.html#catalyst.callbacks.optimizer.OptimizerCallback) in case you don't need AMP.

Another way to reduce memory usage is to disable the calculation of the gradients:
```python
model = ESREncoder(num_basic_blocks=16)
model.load_state_dict(torch.load("path/to/the/model.pth"))
model.eval()

with torch.no_grad():
    out_data = model(data)
```

**Comment by man-sean** [comment]
Oh, I see. Thanks!
